lib/vagrant/util/install_cli_autocomplete.rb

```
@@ -0,0 +1,119 @@
+            f.write(@string_insert)
+            f.write("\n")
+            f.write(@append_string)
+            f.write("\n")
+          end
+        end
+        return path
+      end
+    end
+
+    # Install autocomplete script to zsh config located as .zshrc
+    class InstallZSHShellConfig < InstallShellConfig
+      def initialize()
+        prepend_string = "# >>>> Vagrant zsh completion (start)".freeze
+        string_insert = """fpath=(#{File.join(Vagrant.source_root, "contrib", "zsh")} $fpath)\ncompinit""".freeze
+        append_string = "# <<<<  Vagrant zsh completion (end)".freeze
+        config_paths = [".zshrc".freeze].freeze
+        super(prepend_string, string_insert, append_string, config_paths)
+      end
+    end
+
+    # Install autocomplete script to bash config located as .bashrc or .bash_profile
+    class InstallBashShellConfig < InstallShellConfig
+      def initialize()
+        prepend_string = "# >>>> Vagrant bash completion (start)".freeze
+        string_insert = ". #{File.join(Vagrant.source_root, 'contrib', 'bash', 'completion.sh')}".freeze
+        append_string = "# <<<<  Vagrant bash completion (end)".freeze
+        config_paths = [".bashrc".freeze, ".bash_profile".freeze].freeze
+        super(prepend_string, string_insert, append_string, config_paths)
+      end
+    end
+
+    # Install autocomplete script for supported shells
+    class InstallCLIAutocomplete
+      SUPPORTED_SHELLS = {
+        "zsh" => Vagrant::Util::InstallZSHShellConfig.new(),
+        "bash" => Vagrant::Util::InstallBashShellConfig.new()
+      }
+
+      def self.install(shells=[])
+        shells = shells.length == 0 ? SUPPORTED_SHELLS.keys() : shells
+        home = Dir.home
+        written_paths = []
+        SUPPORTED_SHELLS.each do |k, shell|
+          if shells.include?(k)
+            p = shell.install(home)
+            if p
+              written_paths.push(p)
+            end
+          end
+        end
+        return written_paths
```

This can be compacted down to:

```ruby
shells.map do |shell|
  next if !SUPPORTED_SHELLS[shell]
  SUPPORTED_SHELLS[shell].install(home)
end.compact
```

I'm also wondering if instead of silently ignoring requested shells that are not supported if we should be raising an `ArgumentError`.

---

I think that sounds good!